### DISCUSSION TITLE
non-schema metadata search and enable_dynamic_field

### DISCUSSION THREAD

**Discussion by liuyxme** [milvus-io/milvus#37379] [Q&A and General discussion]
I setup my milvus collection enable_dynamic_field to true, and when I embedded my doc, I customized doc chunk metadata and included some non-schema field. When I use milvus collection query later to get document and include my non-schema field in the output_fields. it didn't show in the returned document attributes. Any idea?

**Comment by xiaofan-luan** [comment]
> I setup my milvus collection enable_dynamic_field to true, and when I embedded my doc, I customized doc chunk metadata and included some non-schema field. When I use milvus collection query later to get document and include my non-schema field in the output_fields. it didn't show in the returned document attributes. Any idea?

did you specify output field when you do query search?  like
res = client.search(
    collection_name="test_collection",
    data=query_vectors,
    filter='color["label"] in ["red"]',
    search_params={
        "metric_type": "L2",
        "params": {"nprobe": 16}
    },
    output_fields=["id", "color"],
    limit=3
)

**Comment by liuyxme** [comment]
I'm using collection.query. like below, "document_type" is non-schema field

> all_documents = collection.query(expr="document_type == 'some_type'",
> output_fields=output_fields)

And there is no document returned

**Reply by yhmo** [reply]
I guess your collection's consistency level is not "Strong".

In milvus, consistency level control the data visibility for search/query: https://milvus.io/docs/consistency.md#Consistency-levels
When you call insert() to insert some data, the data is firstly passed to Pulsar/Kafka as write-ahead-log. Once the data is passed to Pulsar/Kafka, insert() interface returns, the data is still waiting for milvus data node to consume. Only after the data node consumes the data and persist the data to minio/S3, the data is visible to search/query.

For "Bounded", "Eventually" levels, data is not promised to be visible immediately after it is inserted.
With "Strong" level, the query node will wait a while for the data to be consumed and visible so that you can query it,

An example of using enable_dynamic_field and query non-schema metadata. Add consistency_level="Strong" for search() and you will see the documents.
```
from pymilvus import MilvusClient, DataType
import random

collection_name = "test"
dim = 128


if __name__ == '__main__':
    milvus_client = MilvusClient("http://localhost:19530")

    schema = milvus_client.create_schema(enable_dynamic_field=True)
    schema.add_field("id", DataType.INT64, is_primary=True, auto_id=False)
    schema.add_field("vector", DataType.FLOAT_VECTOR, dim=dim)

    index_params = milvus_client.prepare_index_params()
    index_params.add_index(field_name="vector", index_type="FLAT", metric_type="L2")

    milvus_client.create_collection(collection_name=collection_name,
                                    schema=schema,
                                    index_params=index_params,
                                    consistency_level="Bounded")

    rows = [
        {"id": 1, "vector": [random.random() for _ in range(dim)], "document_type": "text"},
        {"id": 2, "vector": [random.random() for _ in range(dim)], "document_type": "picture"},
        {"id": 3, "vector":[random.random() for _ in range(dim)], "document_type": "video"},
    ]
    milvus_client.insert(collection_name, rows)

    all_documents = milvus_client.query(collection_name=collection_name,
                                        filter="document_type == 'picture'",
                                        output_fields=["id", "document_type"],
                                        consistency_level="Strong")
    print(all_documents)
```

**Reply by liuyxme** [reply]
I guess my problem is I was using langchain vectorDatabase.from_document method to insert data to Milvus collection. And even I updated the documents split metadata, those non-schema metadata was not actually stored via langchain to Milvus collection.

**Reply by yhmo** [reply]
In the langchain, the vectorstore.milvus be default creates a collection with enable_dynamic_field=False for you.
https://github.com/langchain-ai/langchain-milvus/blob/1c4cb2b104c6d6146bb5eaebdc18fe87b171ea99/libs/milvus/langchain_milvus/vectorstores/milvus.py#L271

And then infer some fields by the first row you insert:
https://github.com/langchain-ai/langchain-milvus/blob/1c4cb2b104c6d6146bb5eaebdc18fe87b171ea99/libs/milvus/langchain_milvus/vectorstores/milvus.py#L508

If it could not find a matched field, it will skip the metadata:
https://github.com/langchain-ai/langchain-milvus/blob/1c4cb2b104c6d6146bb5eaebdc18fe87b171ea99/libs/milvus/langchain_milvus/vectorstores/milvus.py#L978

So, you can declare a new vectorstore.milvus with enable_dynamic_field=True and try again.

**Reply by liuyxme** [reply]
The milvus collection was created via the pymilvus Collection with enable_dynamic_field=True. 
Now I'm able to query with the non-schema metadata field via collection.query with expr. The change I did was when I inserted document via langChain vectorStore.from_document, I passed enable_dynamic_field=True. 
(Updated) Via LangChain retriever search_kwargs and same expr, also worked. 
Thanks everyone
